modules/Bio/EnsEMBL/Compara/Utils/JIRA.pm

```diff
@@ -22,235 +22,467 @@ package Bio::EnsEMBL::Compara::Utils::JIRA;
+  Description : Submits a post request to the JIRA server that creates a new
+                ticket. Returns the key of the created ticket.
+  Return type : arrayref of strings (JIRA keys)
+  Exceptions  : thrown on invalid $json_input
+
+=cut
+
+sub create_tickets {
+    my $self = shift;
+    my ( $json_input, $issue_type, $priority, $components, $labels, $dry_mode ) = rearrange(
+        [qw(JSON_INPUT ISSUE_TYPE PRIORITY COMPONENTS LABELS DRY_MODE)], @_);
+    # Set default values for optional arguments
+    $issue_type ||= 'Task';
+    $priority ||= 'Major';
+    # Request password (if not available already)
+    my $defined_password = defined $self->{_password};
+    if (! $defined_password) {
+        $self->{_password} = $self->_request_password();
     }
-    
-    if ( $json_hash->{'labels'} ) {
-        foreach my $label ( @{ $json_hash->{'labels'} } ) {
-            $label =~ s/ /_/g; # JIRA doesn't allow whitespace in labels
-            push( @{ $ticket_fields{'labels'} }, $label );
+    # Read tickets from either a JSON formated string or a JSON file path
+    my $json_ticket_list;
+    eval { $json_ticket_list = decode_json($json_input) };
+    # If $@ is not empty, decode_json() has raised an error. Thus, treat
+    # $json_input as a file path.
+    if ($@) {
+        $json_ticket_list = decode_json(slurp($json_input))
+            or die "Could not open file '$json_input' $!";
+    }
+    # Generate the list of JIRA tickets from each JSON hash
+    my $jira_tickets = ();
+    foreach my $json_ticket ( @$json_ticket_list ) {
+        push @$jira_tickets,
+             $self->_json_to_jira($json_ticket, $issue_type, $priority, $components, $labels);
+        if ($json_ticket->{subtasks}) {
+            foreach my $json_subtask ( @{$json_ticket->{subtasks}} ) {
+                push @{$jira_tickets->[-1]->{subtasks}},
+                     $self->_json_to_jira($json_subtask, 'Sub-task', $priority, $components, $labels);
+            }
         }
     }
-
-    if (my $name_on_graph = $json_hash->{'name_on_graph'}) {
-        $name_on_graph =~ s/ /_/g;  # JIRA doesn't allow whitespace in labels
-        push( @{ $ticket_fields{'labels'} }, "Graph:$name_on_graph" );
+    # Log all generated JIRA tickets
+    $self->{_logger}->info(Dumper($jira_tickets) . "\n");
+    my $ticket_key_list = ();
+    if (! $dry_mode) {
+        # JQL queries require whitespaces to be in their Unicode equivalent
+        my $fixVersion = 'Release\u0020' . $self->{_release};
+        # Search for all available tickets in the JIRA server corresponding to
+        # the defined project, release version and division
+        my $jql = sprintf('project=%s AND fixVersion=%s', $self->{_project}, $fixVersion);
+        if ($self->{_division}) {
+            $jql .= sprintf(' AND cf[11130]=%s', $self->{_division});
+        } else {
+            $jql .= ' AND cf[11130] IS EMPTY';
+        }
+        my $division_tickets = $self->fetch_tickets($jql);
```

Comment:
I'm interested in your definition of dry_mode. Seems like you're defining it as "we don't touch JIRA at all". I would consider it to mean "we don't push anything to JIRA". I think it might be useful to perform this check for existing tickets as part of dry_mode, to _really_ show which tickets we plan to create.